add Gmsh mesh file parser

j8xixo12 opened the pull request:
In this commit added mesh file io for modmesh, it lets modmesh has capabilty to read third-party mesh generator generated file and convert the content to a data structure that allow modmesh can use it to do some calculation.

Currently mesh io only support Gmsh format.

j8xixo12 committed:
add Gmsh mesh file parser

In this commit added mesh file io for modmesh, it lets modmesh has
capabilty to read third-party mesh generator generated file and convert
the content to a data structure that allow modmesh can use it to do some
calculation.

Currently mesh io only support Gmsh format.

j8xixo12 commented on modmesh/app/sample_mesh.py:
```diff
@@ -83,6 +83,25 @@ def help_other(set_command=False):
         view.mgr.pycon.command = cmd.strip()
 
 
+def help_mesh_viewer(path, set_command=False):
```
I also added a sample command in sample_mesh app for, once mesh viewer app is completed, then this sample command can be removed.

yungyuc commented:
This PR contains almost 1000 lines of diff and will not be easy to review.  Would it be possible to break it into smaller pieces?  For review efficiency a batch of review should be around 300 lines of diff.

yungyuc commented on modmesh/core.py:
```diff
@@ -64,6 +64,7 @@
     'METAL_BUILT',
     'metal_running',
     'HAS_VIEW',
+    'Gmsh',
```
This is a class and should be put before constants, ordered in the same convention as other classes or functions.

yungyuc commented on cpp/modmesh/io/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+
```
`io` the name conflict with Python `io` module.  It is unnecessary ambiguity.  Please rename this directory as `inout`.

j8xixo12 replied:
I renamed the module name to `inout` in https://github.com/solvcon/modmesh/pull/233/commits/7432534167ee1da998cfb1a7024ed5e8f52f869f

yungyuc replied:
Thanks

yungyuc commented on cpp/modmesh/io/gmsh.hpp:
```diff
@@ -0,0 +1,397 @@
+
+namespace modmesh
+{
+namespace IO
```
namespace should not use uppercase, and it should not have the unnecessary ambiguity to the Python module named `io`. 

In addition, `IO` is a pretty short name and may conflict with symbols in foreign code base.

Please rename it to `inout`.

j8xixo12 replied:
I renamed the namespace to `inout` in https://github.com/solvcon/modmesh/pull/233/commits/7432534167ee1da998cfb1a7024ed5e8f52f869f too.

yungyuc commented on cpp/modmesh/io/gmsh.hpp:
```diff
@@ -0,0 +1,397 @@
+{
+namespace IO
+{
+namespace Gmsh
```
Please do not use uppercase in namespace. Beside, I don't think it is necessary to have 3 levels of namespace. Gmsh is just one of the format. Please remove the namespace.

j8xixo12 replied:
I have removed namespace `Gmsh` in  https://github.com/solvcon/modmesh/pull/233/commits/7432534167ee1da998cfb1a7024ed5e8f52f869f

yungyuc commented on cpp/modmesh/io/gmsh.hpp:
```diff
@@ -0,0 +1,397 @@
+{
+namespace Gmsh
+{
+struct ElementDef
```
What about naming it like `GmshElementDef`?

j8xixo12 replied:
I renamed `ElementDef` to `GmshElementDef` in https://github.com/solvcon/modmesh/pull/233/commits/7432534167ee1da998cfb1a7024ed5e8f52f869f

yungyuc commented on cpp/modmesh/io/gmsh.hpp:
```diff
@@ -0,0 +1,397 @@
+    small_vector<uint8_t> m_mmcl; /* modmesh cell order  */
+}; /* end struct ElementDef */
+
+inline ElementDef ElementDef::by_id(uint16_t id)
```
This is not a template.  Please move it to a cpp file.

j8xixo12 replied:
`GmshElementDef::by_id` is a inline function should I move it to cpp file?
This function will be accessed by gtest test case, if move it to cpp file, gtest will encounter undefined reference.

yungyuc commented on cpp/modmesh/io/gmsh.hpp:
```diff
@@ -0,0 +1,397 @@
+inline ElementDef ElementDef::by_id(uint16_t id)
+{
+#define VEC(...) __VA_ARGS__
+#define MM_DECL_SWITCH_ELEMENT_TYPE(ID, NDIM, NNDS, MMTPN, MMCL) \
```
The use of macro makes the code easy to read.  Thanks.

yungyuc commented on cpp/modmesh/io/gmsh.hpp:
```diff
@@ -0,0 +1,397 @@
+    explicit Gmsh(const std::string & file_path)
+    {
+        bool meta_enter = false, node_enter = false, element_enter = false;
+        // clang-format off
```
You need to turn clang-format back on!!

yungyuc commented on cpp/modmesh/io/gmsh.hpp:
```diff
@@ -0,0 +1,397 @@
+    };
+
+    // Check the finite state machine state transition is valid or not to check msh file format is correct
+    bool _isValidTransition(const std::string s)
```
Do not prepend `_` to any C++ symbols.  You already made them private and mixing the C++ access with Python private convention is a bad idea.

j8xixo12 replied:
I have removed `_` from these private member function.

yungyuc commented on cpp/modmesh/io/gmsh.hpp:
```diff
@@ -0,0 +1,397 @@
+    using real_type = typename number_base::real_type;
+
+public:
+    explicit Gmsh(const std::string & file_path)
```
Could you make it also take a string as data, in addition to a string as path?  Both are useful.

j8xixo12 replied:
I change it to use string as data and python can take care of file stream to `bytes`, even if user is using `Gmsh` in cpp, cpp also can take care of file stream to string.

yungyuc commented on gtests/test_nopython_io.cpp:
```diff
@@ -0,0 +1,350 @@
+#include <modmesh/io/io.hpp>
```
Too long to review.  I trust you have made comprehensive testing.  If not, please speak up.

j8xixo12 replied:
I have tested each cell type conversion between `gmsh` and `modmesh`. Additionally, I also tested if a user inputs a cell type that is not listed in `GmshElementDef`. The expected behavior is that `GmshElement::by_id` should return an empty `GmshElementDef` object.

yungyuc reviewed (changes requested):
Thanks for the great work.  Let's polish the code.

j8xixo12 committed:
code quality enhancement

In this commit, I made the following additions and modifications:

1. Renamed cpp/modmesh/io to cpp/modmesh/inout to prevent
   unnecessary ambiguity.
2. Renamed the namespace IO to inout and removed the Gmsh namespace.
3. Renamed the struct ElementDef to GmshElementDef.
4. Moved non-template functions from gmsh.hpp to gmsh.cpp.
5. The Gmsh constructor now takes a string input as data, and the
   python wrapper also takes py::bytes as input, therefore file IO
   is now taken care of by python.

j8xixo12 committed:
modified gmsh parser and test case for good cross-platform capability

yungyuc commented on cpp/modmesh/inout/gmsh.hpp:
```diff
@@ -0,0 +1,320 @@
+    };
+
+    // Check the finite state machine state transition is valid or not to check msh file format is correct
+    bool isValidTransition(const std::string s)
```
Ah, I forgot to remind that C++ function names should follow snake_case.  You used snake_case in other functions.  This looks like an overlook.  Could you change it and review all in your PR?

j8xixo12 replied:
I fixed this in https://github.com/solvcon/modmesh/pull/233/commits/f7de14ff4eb817cc2c858ee14dfcaa4ec52f2a49
I changed `isValidTransition` and `toblock` to snake case in `Gmsh` class

yungyuc reviewed (changes requested):
It's quite close to completion.  Only some minor formatting issues remain.

j8xixo12 committed:
change gmsh parser function naming style from camel case to snake case

yungyuc commented on cpp/modmesh/inout/pymod/wrap_Gmsh.cpp:
```diff
@@ -0,0 +1,46 @@
+                    [](const py::bytes & data)
+                    { return std::make_shared<inout::Gmsh>(data); }),
+                py::arg("data"))
+            .def("toblock", &wrapped_type::to_block);
```
I guess it's still an overlook here.  It makes sense to align the function names between Python and C++.

yungyuc reviewed (changes requested):
Yet a minor mismatch.  Should be a simple fix 🙂

j8xixo12 committed:
align the Gmsh python api names with c++

yungyuc commented:
Thanks a lot!
